File: Algorithm/Algorithmic Problem Solving Strategies/C++/6.2.cpp
#include <iostream>
#include <vector>

using namespace std;


class Combination
{
public:
	void pick(int n, vector<int>& picked, int toPick);
	void printPicked(vector<int>& picked);
};



void Combination::pick(int n, vector<int>& picked, int toPick) {
	if (toPick == 0) {
		Combination::printPicked(picked);
		return;
	}

	int smallest = picked.empty() ? 0 : picked.back() + 1;

	for (int next = smallest; next < n; ++next) {
		picked.push_back(next);
		pick(n, picked, toPick - 1);
		picked.pop_back();
	}
}


void Combination::printPicked(vector<int>& picked) {
	for (unsigned int i = 0; i < picked.size(); i++) {
		cout << picked[i] << ' ';
	}
	cout << '\n';
}



int main() {
	Combination combination;
	vector<int> a = { 0, 1, 2 };

	combination.pick(10, a, 2);
}

File: Algorithm/Algorithmic Problem Solving Strategies/Go/6.2.go
// 6.2.go
package main

import (
	"fmt"
)

type Combination struct {
}

func (c Combination) pick(n int, picked []*int, toPick int) {
	if toPick == 0 {
		c.printPicked(picked)
		return
	}

	var smallest *int
	if len(picked) == 0 {
		*smallest = 0
	} else {
		*smallest = picked[(len(picked)-1)] + 1
	}

	for next := *smallest; next < n; next++ {
		picked.append(picked, next)
		c.pick(n, picked, toPick)
		picked = picked[:(len(picked) - 2)]
	}
}

func (c Combination) printPicked(picked []*int) {
	for i := 0; i < len(picked); i++ {
		fmt.Println(picked[i])
	}
	fmt.Println('\n')
}

func main() {
	combination := Combination{}
	a := []int{0, 1, 2}

	combination.pick(10, &a, 2)
}
